pkg/scheduler/plugins/gang/gang_cache.go

```diff
@@ -0,0 +1,198 @@
+		gangItems:  make(map[string]*Gang),
+		lock:       new(sync.RWMutex),
+		pluginArgs: &schedulingconfig.GangSchedulingArgs{},
+	}
+}
+
+/*****************Cache BEG*****************/
+func (gangCache *gangCache) createGangToCache(gangId string) *Gang {
+	gangCache.lock.Lock()
+	defer gangCache.lock.Unlock()
+
+	if gang, exist := gangCache.gangItems[gangId]; exist {
+		return gang
+	}
+
+	gang := NewGang(gangId)
+	gangCache.gangItems[gangId] = gang
+	return gang
+}
+
+func (gangCache *gangCache) GetGangFromCache(gangId string) *Gang {
+	gangCache.lock.RLock()
+	defer gangCache.lock.RUnlock()
+
+	return gangCache.gangItems[gangId]
+}
+
+func (gangCache *gangCache) deleteGangFromCache(gangId string) {
+	gangCache.lock.Lock()
+	defer gangCache.lock.Unlock()
+
+	delete(gangCache.gangItems, gangId)
+}
+
+/*****************Cache END*****************/
+
+/******************Pod Event BEG******************/
+func (gangCache *gangCache) OnPodAdd(obj interface{}) {
+	pod, ok := obj.(*v1.Pod)
+	if !ok {
+		return
+	}
+
+	if !IsValidPod(pod) {
+		return
+	}
+	var gang *Gang
+	//the gang is created in either CRD way or Annotation way
+	if _, exist := pod.Labels[v1alpha1.PodGroupLabel]; exist {
+		gangName := pod.Labels[v1alpha1.PodGroupLabel]
+		gangNamespace := pod.Namespace
+		gangId := GetNamespaceSplicingName(gangNamespace, gangName)
+		gang = gangCache.GetGangFromCache(gangId)
+		if gang == nil {
+			klog.Errorf("Didn't find gang from gangCache when pod Added,maybe podGroup is not existed, gangName:%v", gangId)
+			return
+		}
+	} else {
+		gangNamespace := pod.Namespace
+		gangName := pod.Annotations[extension.AnnotationGangName]
+		gangId := GetNamespaceSplicingName(gangNamespace, gangName)
+		gang = gangCache.GetGangFromCache(gangId)
+		if gang == nil {
+			gang = gangCache.createGangToCache(gangId)
+			klog.Infof("Create gang by pod on add, gangName:%v", gangId)
+		}
+		gang.TryInitByPodConfig(pod, gangCache.pluginArgs)
+	}
+	gang.SetChild(pod)
+	gang.SetChildScheduleCycle(pod, 0)
+}
+
+func (gangCache *gangCache) OnPodUpdate(oldObj interface{}, newObj interface{}) {
+}
+
+func (gangCache *gangCache) OnPodDelete(obj interface{}) {
+	pod, ok := obj.(*v1.Pod)
+	if !ok {
+		return
+	}
+
+	if !IsValidPod(pod) {
+		return
+	}
+
+	gangNamespace := pod.Namespace
+	gangName := GetGangNameByPod(pod)
+
+	gangId := GetNamespaceSplicingName(gangNamespace, gangName)
+	gang := gangCache.GetGangFromCache(gangId)
+	if gang == nil {
+		return
+	}
+
+	shouldDeleteGang := gang.deletePod(pod)
+	if shouldDeleteGang {
+		gangCache.deleteGangFromCache(gangId)
+		klog.Infof("Delete gang from gang cache, gangName:%v", gang.Name)
+	}
+}
+
+/******************Pod Event END******************/
+
+/******************GANG Event BEG******************/
+func (gangCache *gangCache) OnGangAdd(obj interface{}) {
+	pg, ok := obj.(*v1alpha1.PodGroup)
+	if !ok {
+		return
+	}
+	if pg.Status.Phase == v1alpha1.PodGroupFinished || pg.Status.Phase == v1alpha1.PodGroupFailed {
```

Comment:
if pg.Status.Phase == v1alpha1.PodGroupFinished || pg.Status.Phase == v1alpha1.PodGroupFailed{}.    I think this can be deleted

Reply:
yes, podGroup will not be handled during the whole scheduling process